### ISSUE TITLE
Agent-ready Qwen3.6-35B recipe recommends nvidia/Qwen3.6-35B-A3B-NVFP4, which silently breaks agentic tool-calling on GB10 (malformed tool-call syntax past a tool-count cliff)

### ISSUE THREAD

**Issue by chiptoe-svg** [NVIDIA/dgx-spark-playbooks#89]
## Summary

The **agent-ready Qwen3.6-35B recipe** (`nvidia/vllm/README.md`, "Run Agent Ready Qwen3.6 35B Model with vLLM") recommends **`nvidia/Qwen3.6-35B-A3B-NVFP4`**, which **silently breaks agentic tool-calling** on DGX Spark (GB10) once the tool surface gets large. The model emits **structurally malformed tool-call syntax** that the parser correctly rejects, so the turn returns `finish_reason: stop` with zero `tool_calls` and the attempt dumped into `content`. It presents as "the model narrates what it intends to do instead of calling tools," which misdirects debugging toward the client, prompt, or parser.

This affects every agentic playbook that reuses the recipe: **`hermes-agent`, `openclaw`, `openshell`, `nemoclaw`**, and the `vllm` playbook itself.

## Environment

- NVIDIA DGX Spark (GB10, sm_121), single node
- vLLM **v0.24.0**, the exact playbook flags: `--reasoning-parser qwen3 --tool-call-parser qwen3_xml --enable-auto-tool-choice`, MTP (`--speculative-config`), `--kv-cache-dtype fp8`, `--attention-backend flashinfer`
- Model: `nvidia/Qwen3.6-35B-A3B-NVFP4` (modelopt)

## Symptom

On instruction-dense requests with a large tool array, the tool-call attempt lands in `content` as malformed XML (should be `<tool_call>\n<function=NAME>\n…`):

```
<tool_call>{"function": "list-gc-catalog-years", "parameters": {}}
{"function": "get-gc-requirement-rules", "parameters": {"year": "2026-2027"}}
</function></tool_call>
```

Three defects: (1) wrong key schema — `{"function":…,"parameters":…}` instead of the `<function=…>` XML the template teaches; (2) multiple calls packed inside one `<tool_call>` block; (3) mismatched `<tool_call>`…`</function>` tags. Other runs produce `{"function:…`, truncated `<tool_call>\n{"`, or garbage close tags (`</anth>`). Short/simple prompts work fine, so it looks like a "decision" failure rather than a generation defect.

## Trigger — a sharp tool-count cliff, and it is NOT size

Same body, only the `tools` array varied, 5 runs each:

| tools | result |
|---|---|
| 20 – 45 | **5/5 clean** |
| **46** | **0/5 (malformed)** |

- **Not size:** padding a passing 40-tool set to 8k / 9k / 10k tokens still returns 5/5. The 46-tool set fails at *less* total weight than the passing padded sets.
- **Order-sensitive:** moving one tool to the front of the failing 46-tool array → 5/5; dropping any one tool → 5/5.
- **Not temperature** (~1 success in 30 across 0.0–1.5) and **not `enable_thinking`** (fails both ways).

## Root cause — isolated by a same-box quant bisect

Identical request body, same vLLM / template / flags, **only the quant varied**:

| quant | tool-calling |
|---|---|
| `nvidia/Qwen3.6-35B-A3B-NVFP4` (modelopt) | **0/5 malformed** |
| `Qwen/Qwen3.6-35B-A3B-FP8` | 5/5 clean |
| `unsloth/Qwen3.6-35B-A3B-NVFP4` (compressed-tensors, dynamic) | 5/5 clean |

So it's the **quant**, not the base model, vLLM, or parser. Mechanistically, the modelopt checkpoint quantizes nearly everything to 4-bit (its `quantization_config` spares only ~2 layers), while unsloth's NVFP4 keeps **311 layers at higher precision** plus a genuine 8-bit/4-bit mix. Aggressive uniform 4-bit appears to flatten the logits that maintain valid structured-output (JSON/XML-tag) tokens once the tool surface is large enough — invisible on short prompts, catastrophic for agents.

## Impact

`nvidia/Qwen3.6-35B-A3B-NVFP4` is the model the playbook presents as the **agent-ready** choice, so anyone building tool-using/agentic workloads on Spark by following the guide hits this — and it fails **silently** (looks like the model declining to call tools), which we can attest sends people down long dead-end debugging paths.

## Recommended action

- **Change the agent-ready recipe's model to `unsloth/Qwen3.6-35B-A3B-NVFP4`** (dynamic quant — clean tool-calling, ~same NVFP4-class speed and memory), or at minimum document the defect and point agentic users to it. `Qwen/Qwen3.6-35B-A3B-FP8` is also clean but slower on GB10 (no native FP8 → weight-only dequant).
- Consider re-releasing `nvidia/Qwen3.6-35B-A3B-NVFP4` with a less-aggressive (structure-preserving) quant recipe.

## Related

Likely the same NVFP4 quant-quality family as #76 (`Phi-4-reasoning-plus-NVFP4` emits garbage tokens) — worth a broader look at the modelopt NVFP4 recipe for reasoning/agentic models on GB10.

## Reproduction / diagnostic tip for others

If a reasoning model "narrates instead of tool-calling" on some backend, inspect the **raw** response: `<tool_call>` XML in `content` with `finish_reason: stop` = malformed *generation* (suspect the quant), not a decision or client-parse issue. Bisecting quants on the same box (holding vLLM/template/flags constant) isolates it in one experiment.

**Comment by kristianpaul** [comment]
I've experience tool call corruption  from this model as well

**Comment by chiptoe-svg** [comment]
**Update — root cause found, with fixes (thanks @kristianpaul for confirming).**

We tracked this down, and it's **not the NVFP4 quant** — it's **vLLM speculative decoding (MTP) + prefix caching interacting with Qwen3.6-35B-A3B's hybrid Mamba/Gated-Delta (SSM) architecture**. Same class as vllm-project/vllm#43650 (MTP + prefix caching + Mamba) and #48970 (hybrid-Mamba prefix-cache corruption). Single-shot calls are usually fine; the *multi-step* accumulation with a large tool surface corrupts, because the prefix-cached Mamba state diverges under speculative verification. Signature matches this issue exactly: `finish_reason: stop`, zero `tool_calls`, malformed `<function=…>` text dumped into `content`.

Two fixes validated on GB10 (sm_121):

1. **On vLLM — drop MTP, keep prefix caching.** Remove `--speculative-config` (NextN) but keep `--enable-prefix-caching`. Restored 0 malformed across a multi-step harness; you lose the MTP decode speedup but keep the prefix-cache TTFT win.
2. **Better — serve on SGLang (≥ v0.5.15.post1).** It's Mamba-aware (`MambaRadixCache` hybrid-SSM prefix cache + NextN with dedicated intermediate SSM state caches), so it runs MTP + prefix caching + the Mamba arch *together* cleanly — 0 malformed multi-step **and** you keep the speculative speedup. Parsers: `--tool-call-parser qwen3_coder --reasoning-parser qwen3` (must be `qwen3_coder`, not `qwen25` — 3.6 emits `<function=…><parameter=…>` XML; `qwen25` silently parses zero tool_calls).

So the recipe fix is a serving-config change (or SGLang), not a quant swap. Happy to share the multi-step reproduction harness.

**Comment by eugr** [comment]
@chiptoe-svg - could you also try with http://github.com/eugr/spark-vllm-docker or just pull eugr/spark-vllm:latest from dockerhub?
If using repository, you can run the latest recommended recipe for this model using `./run-recipe.sh --solo qwen3.6-35b-a3b-nvfp4`

The reason why I'm asking is just because the MTP/cache issue has been addressed after vLLM 0.25.1 was released, so you need more recent but tested build.

**Comment by hebo1221** [comment]
I'm investigating this on another single Spark and would like to validate the playbook change against the same failure shape before proposing it. Could you share the multi-step harness or a sanitized 46-tool request fixture you used for the MTP-on / MTP-off comparison?

I'll hold the model, image, parser flags, prefix caching, and request sequence constant and vary only `--speculative-config`. I also plan to record the throughput cost of the safe configuration so the playbook can state the tradeoff rather than just deleting the flag.

**Comment by hebo1221** [comment]
I tested this independently on a single GB10 with the playbook flags and `nvidia/Qwen3.6-35B-A3B-NVFP4`.

- vLLM 0.24.0, MTP + prefix caching: one deterministic multi-step run produced two identical `lookup_metric_23` calls where exactly one was requested (1 failure in 111 attempted calls).
- vLLM 0.24.0, prefix caching without MTP: 120/120 passed.
- vLLM 0.26.0, MTP + prefix caching: the submitted verifier passed 120/120. Metrics confirmed both paths were active: 83,616 prefix-cache hit tokens and 9,495 MTP draft tokens.

The old-build failure is intermittent: later runs on another 0.24.0 process passed, so a single clean run is not proof of safety.

I opened draft PR #92 to pin the recipe to v0.26.0 and add the 46-tool, 12-step smoke test. It keeps MTP enabled on the fixed build and documents removing MTP as the older-build fallback.

**Comment by chiptoe-svg** [comment]
@hebo1221 @eugr — apologies for the long delay in replying. You both asked me for something and I left it hanging while this box was busy elsewhere; that's on me, and thank you for pushing the work forward without me.

### Harness

Here it is, self-contained: https://gist.github.com/chiptoe-svg/fdc3d3d647108fcbe4d775408cb3281a

The schema is **synthetic on purpose** rather than our real one. Two reasons: our production tool surface belongs to a university service and isn't mine to publish, and — more usefully for you — what reproduces this is the *shape* (tool count, description bulk, multi-step accumulation under prefix reuse), not the semantics. So the script has no MCP dependency, no credentials, and no site-specific anything. It generates N synthetic tools with realistic description bulk (46 tools ≈ 19.4K description chars by default), runs a growing multi-step conversation, and scores exactly the playbooks#89 signature: `finish_reason=stop`, zero `tool_calls`, raw `<function=…>` XML in `content`. Exit code 1 on any malformed call, so it drops straight into CI.

```
python3 qwen36_toolcall_repro.py --url .../v1/chat/completions \
    --model qwen3.6-35b-a3b --tools 46 --steps 12 --trials 10
```

Hold everything constant and vary only `--speculative-config` server-side, which is exactly the comparison you described.

### Independent corroboration of your 0.26.0 result

@hebo1221 — I ran that harness against our production config just now and got the same number you did:

```
schema: 46 tools, 19445 description chars
RESULT: 120 calls, 0 malformed, 0 no-call, 102s
PASS
```

Config: stock `vllm/vllm-openai:v0.26.0`, `nvidia/Qwen3.6-35B-A3B-NVFP4`, **MTP `num_speculative_tokens=3` AND `--enable-prefix-caching` both on**, `--kv-cache-dtype fp8`, `--moe-backend marlin`, `--attention-backend flashinfer`, thinking off. So MTP + prefix caching + hybrid-Mamba together, which is the combination that used to corrupt.

That matches our own on-box verification back in July (0 malformed over 57 tool-call turns on 0.26.0, both thinking modes), and it's now been the production config for advising traffic for about six weeks. Your 120/120 and mine are independent — different hosts, different harnesses, same conclusion.

**So I'd support #92 as written.** Pinning the recipe to v0.26.0 with MTP kept on, and documenting MTP-off as the older-build fallback, matches what two of us measure.

### One correction to my earlier comment

I said the parser "must be `qwen3_coder`, not `qwen25`". The `qwen25` part stands, but on current vLLM the `qwen3_coder` vs `qwen3_xml` distinction is moot — in 0.26.0 both names are lazy aliases to the *same* class (`vllm/tool_parsers/__init__.py` maps both to `Qwen3EngineToolParser`). We run `qwen3_xml` in production. Anyone reading my earlier note shouldn't think the choice between those two matters on a recent build; historically it did.

### @eugr — on your build

Thanks for the pointer, and sorry for not answering sooner. We've stayed on stock `v0.26.0` for a reason unrelated to this issue: vllm-project/vllm#37754, the FlashInfer `BatchPrefillWithPagedKVCache` illegal-memory-access on SM121 with MTP. It's rare for us but real — we captured an EngineCore trace plus a kernel Xid 31 during a scripted reproducer, at `num_running_reqs=1`, and it reproduces identically on driver 580.159.03 and 580.173.02. Since that one is still open with no fix on main, upgrading hasn't bought us anything yet.

That said, your nightlies are the most useful early-warning signal we have for it — you track vLLM main with GB10 patches, so if a FlashInfer sm_121 prefill fix lands upstream it'll surface in your changelog well before it reaches a stock release. We check it for exactly that. Happy to run your `qwen3.6-35b-a3b-nvfp4` recipe against this harness on our GB10 if a comparison would be useful to you — just say the word.
